sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1157 @@
+        ----------
+        .. [1] https://www.python.org/dev/peps/pep-0487
+        """
+        try:
+            requests = cls._get_default_requests()
+        except Exception:
+            # if there are any issues in the default values, it will be raised
+            # when ``get_metadata_routing`` is called. Here we are going to
+            # ignore all the issues such as bad defaults etc.
+            super().__init_subclass__(**kwargs)
+            return
+
+        for method in METHODS:
+            mmr = getattr(requests, method)
+            # set ``set_{method}_request``` methods
+            if not len(mmr.requests):
+                continue
+            setattr(
+                cls,
+                f"set_{method}_request",
+                RequestMethod(method, sorted(mmr.requests.keys())),
+            )
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def _get_default_requests(cls):
+        """Collect default request values.
+
+        This method combines the information present in ``metadata_request__*``
+        class attributes.
+        """
+
+        requests = MetadataRequest(owner=cls.__name__)
+
+        # need to go through the MRO since this is a class attribute and
+        # ``vars`` doesn't report the parent class attributes. We go through
+        # the reverse of the MRO since cls is the first in the tuple and object
+        # is the last.
+        defaults = dict()
+        for klass in reversed(inspect.getmro(cls)):
+            klass_defaults = {
+                attr: value
+                for attr, value in vars(klass).items()
+                if "__metadata_request__" in attr
+            }
+            defaults.update(klass_defaults)
+        defaults = dict(sorted(defaults.items()))
+
+        # First take all arguments from the method signatures and have them as
+        # ERROR_IF_PASSED, except X, y, *args, and **kwargs.
+        for method in METHODS:
+            # Here we use `isfunction` instead of `ismethod` because calling `getattr`
+            # on a class instead of an instance returns an unbound function.
+            if not hasattr(cls, method) or not inspect.isfunction(getattr(cls, method)):
+                continue
+            # ignore the first parameter of the method, which is usually "self"
+            params = list(inspect.signature(getattr(cls, method)).parameters.items())[
+                1:
+            ]
+            for pname, param in params:
+                if pname in {"X", "y", "Y"}:
+                    continue
+                if param.kind in {param.VAR_POSITIONAL, param.VAR_KEYWORD}:
+                    continue
+                getattr(requests, method).add_request(
+                    param=pname,
+                    alias=RequestType.ERROR_IF_PASSED,
+                )
+
+        # Then overwrite those defaults with the ones provided in
+        # __metadata_request__* attributes, which are provided in `requests` here.
```

Comment:
I'm not sure what "which are provided in `requests` here" means.

Reply:
Maybe this comment should be `Defaults set in __metadata_request__ attributes take precedence over signature sniffing`